Added Page Object model and exercise in AddEmployee

monicasantander1 opened the pull request:
Selenium Practical Exercise

msantander001 committed:
Added Page Object model and exercise in AddEmployee

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public class LoginPage : BasePageLocal
+    {
+        private readonly By TxtUserName = By.Name("username"); // locate username
```
Remember... we only use ID, CSS selector, and XPath locator types

monicasantander1 replied:
fixed

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,101 @@
+        // 1. Navigate to https://opensource-demo.orangehrmlive.com/web/index.php/auth/login
+        IWebDriver driver = Driver.Value;
+        driver.Manage().Window.Maximize();
+        driver.Navigate().GoToUrl("https://opensource-demo.orangehrmlive.com/web/index.php/auth/login");
```
This already exists in BaseTest.cs

monicasantander1 replied:
fixed

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,101 @@
+
+        // 2. Log in using Username: Admin, Password: admin123 using LoginPage - Page Object Model
+        LoginPage loginPage = new LoginPage(driver);
+        loginPage.Login("Admin", "admin123"); // enter user & password
```
For ease of use and maintenance, all test data (the username and password) should be declared at the top of the test class.

monicasantander1 replied:
fixed

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,101 @@
+
+
+        // 3. Click "PIM" in the left nav using PimEmployee - Page Object Model
+        new WebDriverWait(driver, TimeSpan.FromSeconds(10))
```
There shouldn't be any WebDriverWait, Driver, etc. calls in the test itself.

monicasantander1 replied:
removed

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,101 @@
+        .Until(ExpectedConditions.ElementToBeClickable(By.Name("firstName")));
+
+        string firstName = Utils.GenerateRandomString(6);
+        string lastName = Utils.GenerateRandomString(8);
```
All test data should be at the top of the test class

monicasantander1 replied:
fixed

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,101 @@
+        verifyEmployee.ClickSave(); // click save
+
+        // allow page to redirect cuz sometimes it takes a while
+        Thread.Sleep(7000);
```
Sleep = bad... don't use sleep.

monicasantander1 replied:
added a wait

jbcrowley commented on SeleniumPracticalExercise/PageObjects/PimEmployee.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace SeleniumPracticalExercise.PageObjects
+{
+    public class PimEmployee : BasePageLocal
```
What's a better name for this class? Think about the entirety of the page/section that you are creating a page object for and name it that rather than just the action you are taking right now.

monicasantander1 replied:
fixed
changed PimEmployee to AddEmployee

jbcrowley commented on SeleniumPracticalExercise/PageObjects/PimEmployee.cs:
```diff
@@ -0,0 +1,35 @@
+            Click(PIMLink);
+        }
+
+        public void ClickADD()
```
Rather than two methods... one to click Add and one to enter the names... think of it like a single action... adding a new employee. Create a single new method to do this that contains all actions needed to complete this task. See Login().

monicasantander1 replied:
added them together and renamed it to ClickPIMAdd

jbcrowley replied:
You have ClickPIMAdd() and EnterNames()... what action is the user taking using both of these methods combined? What is the user trying to do?

The user thinks... I need to do X and the steps are:
1. Click PIM
2. Click Add
3. Enter first/last name

What is X? Name your method X.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/PimEmployee.cs:
```diff
@@ -0,0 +1,35 @@
+        private readonly By PIMLink = By.XPath("//span[text()='PIM']"); // locate PIM button\
+        private readonly By AddBtn = By.CssSelector("div.orangehrm-header-container > button"); // locate Add button
+        private readonly By FirstNameTxtBox = By.Name("firstName");  // locate firstname txtbox
+        private readonly By LastNameTxtBox = By.Name("lastName");    // locate lastname txtbox
```
Use only ID, CSS selector, and XPath

monicasantander1 replied:
fixed

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+{
+    public class VerifyEmployee : BasePageLocal
+    {
+        private readonly By EmployeeId = By.CssSelector("div.oxd-grid-2 input");     // EMployee ID # Locator
```
You don't need to comment each locator, etc. The variable name should be good enough to indicate what the variable stores. Please remove these.

monicasantander1 replied:
removed

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+
+        public VerifyEmployee(IWebDriver driver) : base(driver)
+        {
+
```
remove empty space

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+
+        }
+
```
only one empty line between methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            Click(PIMButton);
+        }
+        public string IDNum() // get Emp ID value 
```
add empty line between methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+        {                 
+            return GetValue(EmployeeId); // well return the ID #
+        }
+
```
remove extra line

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+       public void TxtBoxEmpID(string storeEmployeeID) // txtbox Emp ID but need to insert ID
+        {
+            EditBoxSendKeysAndVerify(TxtBoxEmployeeID, storeEmployeeID);
+           
```
remove empty line

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+            string displayId = GetText(DisplayedID);
+            Assert.AreEqual(employeeId, displayId, "Employee ID is not match expected value.");
+        }
+        public void verifyName(string firstName, string lastName)  // to verify name in searhc result
```
add empty line between methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+        private readonly By SubmitButton = By.CssSelector("button[type='submit']");  // Submit Locator
+        private readonly By DisplayedID = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][2]/div");  // search result displayed ID
+        private readonly By FirstNameDisplay = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][3]/div");  //search result displayed firstname
+        private readonly By LastNameDisplay = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][4]/div");   //search result displayed lastname
```
You can do these more simply with a CSS selector

monicasantander1 replied:
not sure how?
I thought you were testing us since it was a table and it took me 2 hours just to come up with those xpaths.... and I was proud of myself 😭

jbcrowley replied:
We went through this last Thu

monicasantander1 replied:
I know you showed me how to do it in CSS but I forgot.
I tried 
$$("div.oxd-table-cell div")
and tried to add the [1] but didn't get anything.

I'll keep trying and come back to it later

jbcrowley replied:
Don't forget to come back to this

=================================================================
Hey @jbcrowley 
I'm adding a comment here since I can't comment in a separate one.

Question on this,
This was in the VerifyEmployee Page Object I created and Renamed it EmployeeInformationPage.
Locators have been moved to the correct Page Object so what would you like me to come back to fix?
@monicasantander1

jbcrowley replied:
Your last comment was that you were still working on converting the XPaths you used to CSS selectors and that you would come back to it later. I hadn't seen any updates so I was just reminding you.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public class LoginPage : BasePageLocal
+    {
+        private readonly By TxtUserName = By.Name("username"); // locate username
```
Remember when we had the discussion where we name locators as "_blahBlahLocator"? You should do that here, e.g. "_txtUserNameLocator". It makes it MUCH easier to distinguish between the locators and the method parameters if you make this change. The variable starting with a "_" is a ReSharper thing... it will flag it if you don't use it... it's an industry standard thing that we never really talked about.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,25 @@
+        public LoginPage(IWebDriver driver) : base(driver)
+        {
+        }
+
```
Only 1 empty line between methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/PimEmployee.cs:
```diff
@@ -0,0 +1,35 @@
+        public PimEmployee(IWebDriver driver) : base(driver)
+        {
+        }
+
```
1 line between methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+using SeleniumExtras.WaitHelpers;
+using SeleniumPracticalExercise.PageObjects.Common;
+using static SeleniumPracticalExercise.PageObjects.Common.BasePageLocal;
+
```
1 line here

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+           
+        } 
+
+        public void SubmitBut() 
```
SubmitButton()

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,69 @@
+        public void verifyEmployeeId(string employeeId) // to verify ID # in search result
+        {
+            string displayId = GetText(DisplayedID);
+            Assert.AreEqual(employeeId, displayId, "Employee ID is not match expected value.");
```
When you are using asserts, it's really helpful to name them according to whether they are actual or expected, e.g. 'actualId' and 'expectedId'... that way you can put them in the correct assert parameter order easily.

jbcrowley reviewed (changes requested):
Overall you did well. Don't worry about the number of comments... I flagged everything I could find. Some are nitpicks, some are things we haven't talked about yet, etc.

Leave my comments in place. If you have questions about a comment or disagree with it, respond to it. NOTE: There are some comments that apply to a pattern, make sure you fix ALL of them ... not just the one I commented on, e.g. WebDriverWait, Driver, etc. in the test case itself and others.

Fix the stuff I brought up and push that to your branch. Let me know when you've fixed everything and I'll come back and review again.

monicasantander1 committed:
JEFF
I have updated everything but one.

The VerifyEmployee POM with the xpath. you said CSS is better but not sure how

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,86 @@
-
-        }
+        var user = "Admin"; // declaring username for login
+        var passwrd = "admin123"; // declaring passowrd for login
```
I'm not sure if this is a misspelling or an attempted abbreviation but just use `password` as the variable name.

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,86 @@
+        Driver.Value.Navigate().GoToUrl("https://opensource-demo.orangehrmlive.com/web/index.php/auth/login");
+
+        // waiting on page to locate username, if I don't use this it wont enter the username
+        new WebDriverWait(Driver.Value, TimeSpan.FromSeconds(10))
```
In my previous comments I mentioned that there should not be any WebDriverWaits in the test itself.

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,86 @@
+
+        // waiting on page to locate username, if I don't use this it wont enter the username
+        new WebDriverWait(Driver.Value, TimeSpan.FromSeconds(10))
+        .Until(ExpectedConditions.ElementToBeClickable(By.Name("username"))).Click();
```
You've got locators in the test itself which is a no no... all locators go in the page object.

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,86 @@
+        loginPage.Login(user, passwrd); // enter user & password
+
+        // 3. Click "PIM" in the left nav using PimEmployee - Page Object Model
+        SeleniumPracticalExercise.PageObjects.AddingEmployee addEmployee = new SeleniumPracticalExercise.PageObjects.AddingEmployee(Driver.Value);  // Page object model  PIMEmployee
```
Remove `SeleniumPracticalExercise.PageObjects.` here and elsewhere. It's not needed and makes the code harder to read.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,22 @@
+        private readonly By TxtUserNameLocator = By.CssSelector("input[name='username']"); 
+        private readonly By TxtPasswordLocator = By.CssSelector("input[name='password']"); 
+        private readonly By BtnSubmitLocator = By.CssSelector("button");  
+        public LoginPage(IWebDriver driver) : base(driver)
```
Empty line between locators and methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,22 @@
+        private readonly By BtnSubmitLocator = By.CssSelector("button");  
+        public LoginPage(IWebDriver driver) : base(driver)
+        {
+
```
Remove empty line here and in all constructors

monicasantander1 replied:
removed

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+
+        }
+        public void Login(string username, string password)
```
Always have 1 empty line between methods

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddingEmployee.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace SeleniumPracticalExercise.PageObjects
+{
+    public class AddingEmployee : BasePageLocal
```
AddEmployee

monicasantander1 replied:
I changed it to AddingEmployee. Probably wrong name now thinking about it. I probably should have done. AddName. But I thought thats what it was AddingEmployee since that was the page?

monicasantander1 replied:
Renamed it to AddEmployeePage

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddingEmployee.cs:
```diff
@@ -0,0 +1,27 @@
+{
+    public class AddingEmployee : BasePageLocal
+    {
+        private readonly By PIMLink = By.XPath("//span[text()='PIM']"); 
```
All your locators should be in the format `_<element name>Locator`, e.g. `_pimLinkLocator`. Please change them everywhere to be consistent.

monicasantander1 replied:
renamed locators to have the word Locator in it

jbcrowley commented on SeleniumPracticalExercise/PageObjects/Common/BasePageLocal.cs:
```diff
@@ -96,6 +98,7 @@ public virtual string GetValue(By locator)
             throw new Exception($"Not able to get 'value' from element <{locator}> within {timeOut}s.");
         }
 
+
```
Remove this extra empty line

monicasantander1 replied:
Look at the image above, I think you are talking about the extra space in line 95. Above the "throw new Exception"
I have now removed.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/Common/BasePageLocal.cs:
```diff
@@ -1,9 +1,11 @@
 using OpenQA.Selenium.Support.UI;
 using SeleniumExtras.WaitHelpers;
 
+
```
Remove this extra empty line

monicasantander1 replied:
removed extra line

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,86 @@
+        // NOTE:
+        // - Use the provided WebDriver methods in BasePageLocal.cs
+        // - Create page objects as needed
+        // - Document all methods using XML documentation, https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/
```
Don't forget this

monicasantander1 replied:
forgot what this was about

jbcrowley replied:
I was referring to the line, "Document all methods using XML documentation." Make sure you do a pass in the entire project and ensure that all methods have XML docs.

monicasantander1 replied:
made sure to do /// to methods so the correct documentation does show.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            Click(PIMButton);
+        }
+        public string IDNum() // get Emp ID value 
```
In general, if you are getting a value from the page, name your method, `GetX()`, e.g. `GetId()`. If you are setting a value, name your method, `SetX()`, e.g. `SetId()`.

monicasantander1 replied:
confused how to do this. I tried this below and I'm getting an error message. 
       public string IDNum()
        {
            get 
            { 
                return EmployeeId; 
            }
            set 
            { 
                EmployeeId = Value; 
            }
            
        }

jbcrowley replied:
The syntax you are using is for "getters" and "setters". That applies to fields on a class. I'm just talking about method naming... any method that gets a value should be named, `GetSomething()` and any method that sets a value should be named, `SetSomething()`.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/VerifyEmployee.cs:
```diff
@@ -0,0 +1,57 @@
+        }
+        public string IDNum() // get Emp ID value 
+        {                 
+            return GetValue(EmployeeId); // well return the ID #
```
Remove end of line comments here and elsewhere. Add XML docs instead.

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,86 @@
-            // - Document all methods using XML documentation, https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/
-
-        }
+        var user = "Admin"; // declaring username for login
```
In general, use comments sparingly. In this case, it should be obvious by looking at the code that you are storing a string in a variable. Save comments to describe code that isn't easy to understand, etc.

jbcrowley replied:
We don't use `var`... use defined types.

jbcrowley replied:
username

jbcrowley reviewed (changes requested):
Overall things are getting better. Most of the comments at this point are styling issues. Make sure you read each comment carefully and apply it not only to the line I've commented on but the entire project.

monicasantander1 committed:
JEFFFFFFF
I made my updated. you can just pass me... hehe

monicasantander1 committed:
Jeff,
I made my update butttt I'm sure I didn't do AddEmployeePage right

monicasantander1 committed:
Hey Jeff,
working on this since I still felt I did it wrong on Friday. I think its better but not 100% there

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -6,19 +6,20 @@ namespace SeleniumPracticalExercise.PageObjects
     public class EmployeeInformationPage : BasePageLocal
     {
-        
+        private readonly By _actualFirstNameLocator = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][3]/div");
```
Remember, the "actual"/"expected" prefix for variables shouldn't be used for locators. It was intended only to be used around Asserts.

monicasantander1 replied:
Renamed them to have ResultsLocator

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -6,19 +6,20 @@ namespace SeleniumPracticalExercise.PageObjects
-        private readonly By _expectedIDLocator = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][2]/div");
-        private readonly By _expectedFirstNameLocator = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][3]/div");
-        private readonly By _expectedLastNameLocator = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][4]/div");
+        private readonly By _employeeIDTxtBoxLocator = By.CssSelector("div.oxd-input-group input.oxd-input"); 
```
We talked in general about not using abbreviations, e.g. "Txt".

In this case, use "Field" instead of "TxtBox"... so "_employeeIdFieldLocator". We haven't talked about that I don't think? It's a automation team preference just to be consistent.

monicasantander1 replied:
rename to have Field instead of textbox

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -6,19 +6,20 @@ namespace SeleniumPracticalExercise.PageObjects
     {
-        
+        private readonly By _actualFirstNameLocator = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][3]/div");
+        private readonly By _actualIDLocator = By.XPath("//div[contains(@class,'oxd-table-row--clickable')]//div[text()=''][2]/div");
```
Remember, variables should have camel case names... even common acronyms, e.g. "ID" should be "Id"... "_actualIdLocator".

monicasantander1 replied:
renamed to idResultsLocator

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -6,19 +6,20 @@ namespace SeleniumPracticalExercise.PageObjects
         }
 
         /// <summary>
+        /// PIM button
```
This XML doc for a method should only describe what the method does. For example, in this case "Clicks the Add button" or whatever. The rest should be removed here and elsewhere.

monicasantander1 replied:
update my XML and was moved to the LeftNavPanel page object model

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -28,36 +29,38 @@ public void AddButton()
 
         /// <summary>
-        ///  txtbox Emp ID and insert stored ID
+        /// this is the second time on page where we...... 
```
Make sure you create your XML doc *after* adding all parameters, etc... pretty much when you're one with it. That way, when you create the XML doc it has all the pieces. In this case you are missing docs for the parameter, storeEmployeeId. To add it, I usually copy out the docs that I have already, delete the rest, and then start a new doc. That should put the skeleton in for each of the parts, then paste in what you had before and fill in the rest.

monicasantander1 replied:
updated my XML in SetEmployeeIdField to have param in it

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -28,36 +29,38 @@ public void AddButton()
+        /// <summary>
+        ///  this is to verify ID # in search result
+        /// </summary>
+        public string verifyEmployeeId()
```
All method names should be Pascal case.

monicasantander1 replied:
updated to VerifyEmployeeId()

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -28,36 +29,38 @@ public void AddButton()
+        ///  this is to verify first name in search result
         /// </summary>
-        public void verifyEmployeeId(string employeeId)
+        public string verifyFirstNameResults()
```
This method name no longer matches what it's doing.

monicasantander1 replied:
renamed it to FirstNameSearchResults 
I left it with verify cause we would be using it to verify the first name but I see now we will use verify in the assert

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -28,36 +29,38 @@ public void AddButton()
+        ///  this is to verify last name in search result
         /// </summary>
-        public void verifyName(string firstName, string lastName)
+        public string verifyLasttNameResults()
```
This method name no longer matches what it's doing.

monicasantander1 replied:
renamed it to LastNameSearchResults 
I left it with verify cause we would be using it to verify the lastname but I see now we will use verify in the assert

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -28,36 +29,38 @@ public void AddButton()
+        /// search for emp id textbox and insert stored ID
+        /// then we click search button
         /// </summary>
         public void SetTxtBoxEmpID(string storeEmployeeID)
```
No abbreviations. Pascal Case. No "TxtBox" unless needed.

monicasantander1 replied:
Renamed it to SetEmployeeIdField

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -28,36 +29,38 @@ public void AddButton()
+        /// <summary>
+        ///  this is to verify ID # in search result
+        /// </summary>
+        public string verifyEmployeeId()
```
This method name no longer matches what it's doing.

monicasantander1 replied:
renamed it to EmployeeIdSearchResults 
I left it with verify cause we would be using it to verify the the employee ID but I see now we will use verify in the assert

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -50,13 +42,15 @@ public void AddEmployeeTest()
+            string actualFirstNameResults = employeeInformationPage.verifyFirstNameResults();
+            string actualLastNameResults = employeeInformationPage.verifyLasttNameResults();
+
+            Assert.AreEqual(storeEmployeeID, actualIdResults, "ID does not match expected value.");
```
I don't think we've talked about this yet but we write assert comments as the action being taken, e.g. "Verify employee ID".

What you have isn't wrong... this is just another case where the automation team has agreed to do something a certain way for consistency.

monicasantander1 replied:
renamed them to have verify in it

jbcrowley replied:
The entire assert comment should be, "Verify employee ID". Remove "does not match expected value." Follow this pattern for the other asserts as well.

monicasantander1 replied:
rename them to....
 Verify Employee ID.
 Verify First Name.
Verify Last Name.

jbcrowley reviewed (changes requested):
Getting there...

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,42 @@
+        /// then Click Save
+        /// did stale so we can wait  for element to save. if not it will not find element later
+        /// </summary>
+        public void EnterNames(string firstname, string lastname)
```
Remember... we use the "Set" prefix for methods that set values, e.g. `SetSomething()`.

monicasantander1 replied:
@jbcrowley 
So in EmployeeInformationPage 
should I rename the VerifyEmployeeId, VerifyFirstNameResults, VerifyLastNameResults
to have Set prefix?
ex: SetVerifyEmployeeId()

jbcrowley replied:
These methods no longer exist but VerifyX() is fine as long as what it's doing is verifying X.

monicasantander1 replied:
I believe the VerifyEmployeeId and others are now in the EmployeeInformationPage and are changed to GetEmployeeIDSearchResults() etc

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,42 @@
+        /// <summary>
+        /// get Emp ID value & returns the ID #
+        /// </summary>
+        public string GetIDNum() 
```
Pascal case, no abbreviations.

monicasantander1 replied:
Renamed to GetIdNum()

jbcrowley replied:
Num is an abbreviation

jbcrowley commented on SeleniumPracticalExercise/PageObjects/Common/BasePageLocal.cs:
```diff
@@ -153,7 +155,7 @@ public virtual IWebElement FindElement(By locator)
         }
 
-        /// <summary>
+        /// **<summary>
```
Not sure where this came from and why but it should be removed.

You'll want to look carefully at each line of your changes before you commit. That way you don't accidentally make changes you didn't intend to, you get a chance to code review your own changes, etc.

monicasantander1 replied:
removed the ** 
I was confused on both being named FindElement but Jeff talked to me about it.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -0,0 +1,66 @@
+        /// Click Add Button
+        /// Enter text in Firstname field and LastName field
+        /// </summary>
+        public void AddButton()
```
This should be `Add()` or `ClickAdd()`. The only time I would specify "Button" in a method name is to distinguish maybe a Add link from an Add button so we'd have, `AddButton()` and `AddLink()`... otherwise just `Add()` is fine.

monicasantander1 replied:
Renamed it to ClickAdd()

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LeftNavPanel.cs:
```diff
@@ -0,0 +1,22 @@
+        /// <summary>
+        /// Click PIM on Left Navigation Panel
+        /// </summary>
+        public void ClickPIM()
```
Pascal case

monicasantander1 replied:
Renamed it to ClickPim()

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,56 @@
+            Driver.Value.Navigate().GoToUrl("https://opensource-demo.orangehrmlive.com/web/index.php/auth/login");
+
+            // 2. Log in using Username: Admin, Password: admin123 using LoginPage - Page Object Model
+            LoginPage loginPage = new LoginPage(Driver.Value);
```
Make this a one-liner... remember we talked about this last Thu? If the `loginPage` variable (or any variable for that matter) isn't used more than once, we generally make it a one-liner. Having said that, if that one line becomes excessively long then two lines are fine. Readability should always trump fewer lines.

monicasantander1 replied:
made it a one-liner
new LoginPage(Driver.Value).Login(username, password);

jbcrowley reviewed (changes requested):
OK... I've cleaned up the old comments and I think all that's left is new/valid comments. Make sure you review all comments and make changes, where applicable, throughout the entire project. We're getting close...

monicasantander1 committed:
Jeff,
I have made my updates. Ready for a review
Thank you

monicasantander1 committed:
I have gone through all the "changes from commit" and commented on every one to make sure we see it and updates have been made.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,41 @@
+        /// then Click Save
+        /// did stale so we can wait for element to save. 
+        /// </summary>
+        public void SetEnterNames(string firstname, string lastname)
```
Set and Enter are redundant... just SetName() would be better

monicasantander1 replied:
Renamed to... 
public void SetNames(string firstname, string lastname)

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,41 @@
+
+        /// <summary>
+        /// Enter Firstname & Lastname
+        /// then Click Save
```
You should do a pass on all of your XML docs. I left a comment earlier about this... they should provide a high level description of what the method does, not a step by step description.

"Enter the new employee's first and last name" or something like that. Short and sweet.

jbcrowley replied:
If you want some examples, look at the docs in BasePageLocal.cs

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,41 @@
+        /// <summary>
+        /// get Employee Id value & returns the Id #
+        /// </summary>
+        public string GetIdNum() 
```
Num is an abbreviation

jbcrowley replied:
Remember... name things using the labels on the page whenever possible. What's the label for this ID field on the page?

monicasantander1 replied:
it display Employee Id so I renamed it to
 GetEmployeeId()

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        ///  Grabbing the Text from the ID in the search results to verify
+        /// </summary>
+        public string EmployeeIdSearchResults()
```
This method is getting something so it should be named...?

monicasantander1 replied:
renamed to GetEmployeeIdSearchResults()

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        ///  Grabbing the text from the first name in search result to verify
+        /// </summary>
+        public string FirstNameSearchResults()
```
This method is getting something so it should be named...?

monicasantander1 replied:
renamed GetFirstNameSearchResults()

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -0,0 +1,63 @@
+        /// <summary>
+        ///  Grabbing the text from the last name in search result to verify
+        /// </summary>
+        public string LastNameSearchResults()
```
This method is getting something so it should be named...?

monicasantander1 replied:
renamed to GetLastNameSearchResults()

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,54 @@
-            // - Create page objects as needed
-            // - Document all methods using XML documentation, https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/
+            string verifyIdResults = employeeInformationPage.EmployeeIdSearchResults();
+            string verifyFirstNameResults = employeeInformationPage.FirstNameSearchResults();
```
Remember we talked about naming variables for asserts as "actual" or "expected"?

monicasantander1 replied:
I 100% did do this. Don't know why it wasn't showing but it renamed them again.

  string actualIdResults = employeeInformationPage.GetEmployeeIdSearchResults();
  string actualFirstNameResults = employeeInformationPage.GetFirstNameSearchResults();
  string actualLastNameResults = employeeInformationPage.GetLastNameSearchResults();

jbcrowley reviewed (changes requested):
This is looking really good... we're just polishing at this point.

I responded to some older comments that aren't resolved. Make sure you go through those too.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,41 @@
+        /// then Click Save
+        /// did stale so we can wait for element to save. 
+        /// </summary>
+        public void SetEnterNames(string firstname, string lastname)
```
You are missing XML doc parameters. Add those here and check all the other docs for missing params.

monicasantander1 replied:
went through all page objects I created and re-did the XML documentation

monicasantander1 commented:
> This is looking really good... we're just polishing at this point.
> 
> I responded to some older comments that aren't resolved. Make sure you go through those too.

monicasantander1 committed:
Jeff,
went through comments and made my updates.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -17,11 +17,11 @@ public AddEmployeePage(IWebDriver driver) : base(driver)
-        public void SetEnterNames(string firstname, string lastname)
+        /// <param name="firstname"></param>
+        /// <param name="lastname"></param>
+        public void SetNames(string firstname, string lastname)
```
I missed this before... 'firstname' should be 'firstName' and so on... camel case for variable names here. Check this everywhere else as well.

jbcrowley replied:
Also... make sure you use CTRL+R, CTRL+R to rename the variable so that all instances, including those in the XML docs, are renamed.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -17,11 +17,11 @@ public AddEmployeePage(IWebDriver driver) : base(driver)
+        /// Enter First and Last name
         /// </summary>
-        public void SetEnterNames(string firstname, string lastname)
+        /// <param name="firstname"></param>
```
These docs are empty. You should describe each parameter... I saw this elsewhere as well so make sure you go back and fill all of those in as well. Examples can be found in BasePageLocal.cs.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -31,9 +31,10 @@ public void SetEnterNames(string firstname, string lastname)
 
         /// <summary>
-        /// get Employee Id value & returns the Id #
+        /// get Employee Id value & returns Id #
```
This can be simply, "Returns the employee ID." Don't worry about the "get" portion of the description... just describe the overall goal of the method as simply as you can.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,8 +26,7 @@ public void ClickAdd()
         /// <summary>
-        /// search for employee ID field and insert stored ID
-        /// then we click search button
+        /// search for employee ID textbox field and insert ID
```
This can be simply, "Sets the employee ID to the provided value" or something like that... See FindElement() or SendKeys() in BasePageLocal.cs.

jbcrowley replied:
"Sets the employee Id to employeeId" is too literal. We shouldn't need to restate exactly what the code does because the code is there for them to see if they need that level of detail.

It should be, "Sets the employee ID to the provided value" or "Sets the employee ID."

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -37,25 +36,28 @@ public void SetEmployeeIdField(string storeEmployeeID)
 
         /// <summary>
-        ///  Grabbing the Text from the ID in the search results to verify
+        /// Grabbing the text from ID in the search results
```
"Gets the employee ID from search results"... follow this pattern in the other two methods below.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -14,8 +14,10 @@ public LoginPage(IWebDriver driver) : base(driver)
 
         /// <summary>
-        /// Enter (SendKeys) for username and password then Click Submit button
+        /// Enter username and password 
```
It does more than just enters the username and password... it logs in. Should be something more like, "Logs in using the provided username and password."

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -31,9 +31,10 @@ public void SetEnterNames(string firstname, string lastname)
+        /// get Employee Id value & returns Id #
         /// </summary>
-        public string GetIdNum() 
+        /// <returns></returns>
```
Make sure you are filling in the returns field as well

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -0,0 +1,63 @@
+        /// then we click search button
+        /// </summary>
+        /// <param name="storeEmployeeID"></param>
+        public void SetEmployeeIdField(string storeEmployeeID)
```
This method name no longer matches what it's doing.

monicasantander1 replied:
I renamed it to SetEmployeeId() 
but I am a little confused why you said the method name does not match what its doing 
since it is dealing with the EmployeeId
Was it that word Field I added, since we are grabbing the id and setting it to storeEmployeeId, we don't need the word Field?

jbcrowley replied:
This method is searching for an employee. Don't look at the individual steps, look at what it's doing holistically.

The method name could be 'EmployeeSearch' or 'EmployeeSearchById` or if employee search is the only search type possible it could be shortened to 'Search' or 'SearchById'. I would go with 'SearchById' because employee search is the only search possible. It's also on the employee information page so employee search is at least implied. Also, I can imagine that there will likely be other types of employee searches using other fields so 'SearchById' already makes room for 'SearchByName' or 'SearchByTeam' or whatever other methods there might be as well as the generic 'Search' which would take all fields as optional parameters, letting you use any combination of search parameters.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -0,0 +1,63 @@
+        /// then we click search button
+        /// </summary>
+        /// <param name="storeEmployeeID"></param>
+        public void SetEmployeeIdField(string storeEmployeeID)
```
This parameter should just be 'employeeId'. We want to be consistent with naming of "things"... it's called the employee ID in the XML doc, the name of the method, the locator, and on the web page.

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,54 @@
+            string firstName = Utils.GenerateRandomString(6); 
+            string lastName = Utils.GenerateRandomString(8);
+
             // Steps to automate:
```
Don't forget to remove all of these commented steps for the test case as we discussed before.

monicasantander1 replied:
painful for me to do but I did it

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -1,31 +1,54 @@
+            // 6. Get the Employee Id for use later - using VerifyEmployee POM
             // 7. Click Save
+            AddEmployeePage addEmployeePage = new AddEmployeePage(Driver.Value);
+            string storeEmployeeID = addEmployeePage.GetEmployeeId();
```
camel case... storeEmployeeId

jbcrowley reviewed (changes requested):
Now that everything is pretty clean... I went back and did a full pass again and found a few more issues. I cleaned up all the old comments so this should be it. You're doing well... keep going.

monicasantander1 committed:
Made my updates.
Ready for review

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -19,21 +19,21 @@ public AddEmployeePage(IWebDriver driver) : base(driver)
-        /// <param name="firstname"></param>
-        /// <param name="lastname"></param>
-        public void SetNames(string firstname, string lastname)
+        /// <param name="firstName">we store firstName.</param>
```
>Please< look at the examples in BasePageLocal.cs. This and others need improvement.

monicasantander1 replied:
I looked at BasePageLocal but not sure if I am overthinking it.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -19,21 +19,21 @@ public AddEmployeePage(IWebDriver driver) : base(driver)
+        /// Returns the employee ID
         /// </summary>
-        /// <returns></returns>
+        /// <returns>The Id value contained in the employee id field./</returns>
```
What does it return? "The employee ID". I think you're overthinking these or something? Just make it simple, less verbose.

monicasantander1 replied:
@jbcrowley 
I think I am using these as comments and that's why I'm getting confused and the same with the param.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,37 +26,37 @@ public void ClickAdd()
 
         /// <summary>
-        /// search for employee ID textbox field and insert ID
+        /// Sets the employee Id to employeeId
```
Sets the employee ID

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,37 +26,37 @@ public void ClickAdd()
         /// </summary>
-        /// <param name="storeEmployeeID"></param>
-        public void SetEmployeeIdField(string storeEmployeeID)
+        /// <param name="employeeId"> We are storing the locator id here. </param>
```
This parameter does not store the locator ID.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,37 +26,37 @@ public void ClickAdd()
 
         /// <summary>
-        /// Grabbing the text from ID in the search results
+        /// Gets the employee ID from search results
```
"**Returns** the employee ID from search results"

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,37 +26,37 @@ public void ClickAdd()
+        /// Gets the employee ID from search results
         /// </summary>
-        /// <returns></returns>
+        /// <returns>The text in the Employeeid field.</returns>
```
You're being too literal? "The employee ID"

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -14,13 +14,13 @@ public LoginPage(IWebDriver driver) : base(driver)
-        /// <param name="username"></param>
-        /// <param name="password"></param>
-        public void Login(string username, string password)
+        /// <param name="userName">we store username.</param>
```
This variable should be 'username' because it's 'username' not 'user name'. Make sense? Pascal case capitalizes the first letter of each word. 'username' is one word so no letters are capitalized.

jbcrowley replied:
"we store username"? What does that even mean? Please look at example in BasePageLocal.cs

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -9,45 +9,34 @@ class AddEmployee : BaseTestLocal
         public void AddEmployeeTest()
         {
-            string username = "Admin"; 
+            string userName = "Admin"; 
```
username

jbcrowley commented on SeleniumPracticalExercise/TestCases/AddEmployee.cs:
```diff
@@ -9,45 +9,34 @@ class AddEmployee : BaseTestLocal
-            // 7. Click Save
             AddEmployeePage addEmployeePage = new AddEmployeePage(Driver.Value);
-            string storeEmployeeID = addEmployeePage.GetEmployeeId();
+            string storeEmployeeId = addEmployeePage.GetEmployeeId();
```
employeeId... consistency is key. There's no reference to 'store' anywhere.

jbcrowley reviewed (changes requested):
Please slow down and take a few minutes when updating the XML docs. These are important and will be used by everyone to better understand the code, parameters, and returns for each method.

The summary should be an English sentence with nouns, verbs, and proper punctuation and capitalization.

Each parameter summary should answer the question, "What does this variable store?" The answer can be simple like, "The username" or more complex when needed like, "A collection of customer IDs". We don't reference data types because that's already provided in the method signature.

Each return summary should answer the question, "What does this method return?"

We're talking ELI5, executive summaries... no gory details. If they need the details, they can read the code.

monicasantander1 committed:
Jeff,
I have made my updates and I am ready for a review.
If I am still getting Param wrong, Can we meet?
I think I am overthinking it.
--- Thank you again

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -19,8 +19,8 @@ public AddEmployeePage(IWebDriver driver) : base(driver)
         /// </summary>
-        /// <param name="firstName">we store firstName.</param>
-        /// <param name="lastName">we store lastName.</param>
+        /// <param name="firstName">The firstName used to find element .</param>
```
'used to find element' is internal details that the user doesn't need to know. Something simple like, "The employee's first name" is enough. Same for last name, below, and anywhere else.

jbcrowley replied:
Also, don't refer to the variable name. The user already has that. So use 'first name' instead of 'firstName', when needed.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,37 +26,37 @@ public void ClickAdd()
 
         /// <summary>
-        /// Sets the employee Id to employeeId
+        /// Sets the employee ID to the provided value
```
We're not setting a value here... just filling in the search fields. Something like, "Searches for an employee by employee ID" is good.

jbcrowley commented on SeleniumPracticalExercise/PageObjects/EmployeeInformationPage.cs:
```diff
@@ -26,37 +26,37 @@ public void ClickAdd()
+        /// Returns the employee ID from search results
         /// </summary>
-        /// <returns>The text in the Employeeid field.</returns>
+        /// <returns>The Employeeid Id.</returns>
```
Typo

jbcrowley commented on SeleniumPracticalExercise/PageObjects/LoginPage.cs:
```diff
@@ -16,11 +16,11 @@ public LoginPage(IWebDriver driver) : base(driver)
-        /// <param name="userName">we store username.</param>
-        /// <param name="password"> we store password.</param>
-        public void Login(string userName, string password)
+        /// <param name="username">username used to find element.</param>
```
Can be simply, "The employee's username"

jbcrowley reviewed (changes requested):
You are probably frustrated by now... don't be, you are doing well. We're cramming a lot of training in this PR, some of which wasn't covered before and probably should be added to the curriculum.

Naming is hard. XML docs are hard. You will get better at both with experience. You can always refer back to BasePageLocal.cs for examples. When I first started writing my docs, I used .NET as my examples and tried to emulate them. I still struggle with naming and docs from time to time... don't worry about it.

We're getting closer... just a few comments left. Do a final pass on all your docs and make sure they are good, consistent, etc.

monicasantander1 committed:
Ready for my FINAL REVIEW...
hopefully?

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -0,0 +1,42 @@
+        /// </summary>
+        /// <param name="firstName">The employee's first name.</param>
+        /// <param name="lastName">The employee's last name.</param>
+        public void SetNames(string firstName, string lastName)
```
Sorry, I missed this earlier... this method doesn't just set names... what is it doing?

monicasantander1 replied:
GetNames

jbcrowley replied:
Where does this method GET anything? It's definitely not returning anything. What does it do?

jbcrowley reviewed (changes requested):
Missed one last thing... everything else looks great.

monicasantander1 committed:
Changed SetNames to GetNames

monicasantander1 committed:
AddEmployee() update

jbcrowley commented on SeleniumPracticalExercise/PageObjects/AddEmployeePage.cs:
```diff
@@ -17,11 +17,11 @@ public AddEmployeePage(IWebDriver driver) : base(driver)
 
         /// <summary>
-        /// Enter First and Last name
+        /// Adding an Employee's name
```
It doesn't just add an employee's name, it adds an employee. It just so happens that the only fields we're filling in for this method are first and last name. "Adds an employee" would be more accurate.

jbcrowley reviewed (changes requested):
One last fix

monicasantander1 committed:
Changed...Adds an Employee
